### PR TITLE
[FIX] sale: fix search "invoices is not set" for sale.order

### PR DESCRIPTION
**Pull request by yelizariev**
Previous attempt #70343 of fixing was incorrect. Check code comments for
details.

---

opw-2516124

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by yelizariev** [comment]
@nle-odoo

**Review by Feyensv — COMMENTED** [review]
Could you add a test to cover this specific search method please ?
It's a tricky case, left untested long enough and it could be useful to ensure it's not broken by orm changes in 13+ (a lot of work was done in domain and queries in latest versions).

**Review by nle-odoo — COMMENTED** [review]
ok, it will create queries will all invoiced order (when the query could be a `limit=1`) but that might be premature optimization since anyway this was something that was returning no result before the original change.

robodoo delegate+

**Comment by yelizariev** [comment]
@Feyensv done

File: addons/sale/tests/test_sale_to_invoice.py [hunk 1]
```diff
@@ -68,19 +68,40 @@ def setUpClass(cls):
             'default_journal_id': cls.journal_sale.id,
         }
 
+    def check_has_invoices(self, has_invoices):
```
Comment (Feyensv) [inline addons/sale/tests/test_sale_to_invoice.py:71]:
```suggestion
    def _check_has_invoices(self, order, has_invoices):
```
Give the order as a param, it'll be more generic and easy to use for future/other tests :)

File: addons/sale/tests/test_sale_to_invoice.py [hunk 2]
```diff
@@ -68,19 +68,40 @@ def setUpClass(cls):
...
+        """Test "Invoices is not set" query when some lines are not invoiced"""
+        self.sol_prod_order.product_uom_qty = 0
+        # Confirm the SO
+        self.sale_order.action_confirm()
+        self.check_has_invoices(False)
+        # Let's do an invoice with invoiceable lines
+        payment = self.env['sale.advance.payment.inv'].with_context(self.context).create({
+            'advance_payment_method': 'all'
+        })
+        payment.create_invoices()
```
Comment (Feyensv) [inline addons/sale/tests/test_sale_to_invoice.py:87]: Why do you need to go through the wizard ?
`self.sale_order.action_invoice_create()` should be enough :thinking:

File: addons/sale/tests/test_sale_to_invoice.py [hunk 3]
```diff
@@ -68,19 +68,40 @@ def setUpClass(cls):
             'default_journal_id': cls.journal_sale.id,
         }
 
+    def check_has_invoices(self, has_invoices):
+        """Check query "Invoices is not set" """
+        domain = [('id', '=', self.sale_order.id), ('invoice_ids', '=', False)]
+        result = self.env['sale.order'].search(domain)
+        self.assertEqual(not result, has_invoices, "Domain [('invoice_ids', '=', False)] doesn't return expected result")
+
+    def test_searching_has_invices(self):
```
Comment (Feyensv) [inline addons/sale/tests/test_sale_to_invoice.py:77]:
```suggestion
    def test_searching_has_invoices(self):
```

**Review by Feyensv — COMMENTED** [review]
Some little comments ;)
If you don't mind, could you also test other domains than `[('invoice_ids', '=', False)]` ?

**Comment by yelizariev** [comment]
> If you don't mind, could you also test other domains than `[('invoice_ids', '=', False)]` ?

Done!

```py

self._check_order_search(self.sale_order, [('invoice_ids', 'in', invoice_ids)], self.sale_order)  
self._check_order_search(self.sale_order, [('invoice_ids', '=', False)], self.env['sale.order'])       
self._check_order_search(self.sale_order, [('invoice_ids', '!=', False)], self.sale_order)
```

**Review by Feyensv — APPROVED** [review]

**Comment by yelizariev** [comment]
robodoo r+

**Pull request closed without merging**